### ISSUE TITLE
Failed to rollback writer on close (NPE)

### ISSUE THREAD

**Issue by skundrik** [elastic/elasticsearch#15819]
Got this NPE today on 1 of our data nodes which then proceeded to recovery.
ES 2.1, OS Win2008 R2
```
[2016-01-07 11:17:22,718][WARN ][index.engine             ] [data-nd2] [.marvel-es-2016.01.07][0] failed to rollback writer on close
java.lang.NullPointerException
[2016-01-07 11:17:22,723][WARN ][index.engine             ] [data-nd2] [.marvel-es-2016.01.07][0] failed engine [refresh failed]
java.lang.NullPointerException
[2016-01-07 11:17:22,724][WARN ][index.shard              ] [data-nd2] [.marvel-es-2016.01.07][0] Failed to perform scheduled engine refresh
[.marvel-es-2016.01.07][[.marvel-es-2016.01.07][0]] RefreshFailedEngineException[Refresh failed]; nested: NullPointerException;
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:686)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:615)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1255)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
[2016-01-07 11:17:22,724][WARN ][indices.cluster          ] [data-nd2] [[.marvel-es-2016.01.07][0]] marking and sending shard failed due to [engine failure, reason [refresh failed]]
java.lang.NullPointerException
```

**Comment by s1monw** [comment]
I don't have any idea how that NullPointerException showed up without a stacktrace to be honest! @rmuir any idea?

**Comment by skundrik** [comment]
I came across this article which talks about JVM optimisation of NullPointerException. It might be worth adding this as a standard start-up option.

http://jawspeak.com/2010/05/26/hotspot-caused-exceptions-to-lose-their-stack-traces-in-production-and-the-fix/

**Comment by skundrik** [comment]
Based on the above  I went to see if I can find the first occurrence of it and came up with this.
```
[2015-12-23 11:47:49,949][WARN ][index.engine             ] [data-nd2] [logstash-2015.12.23-v1][1] failed to rollback writer on close
java.lang.NullPointerException
	at org.elasticsearch.indices.cache.query.IndicesQueryCache$1.onDocIdSetEviction(IndicesQueryCache.java:158)
	at org.apache.lucene.search.LRUQueryCache.clearCoreCacheKey(LRUQueryCache.java:313)
	at org.apache.lucene.search.LRUQueryCache$1.onClose(LRUQueryCache.java:276)
	at org.apache.lucene.index.SegmentCoreReaders.notifyCoreClosedListeners(SegmentCoreReaders.java:168)
	at org.apache.lucene.index.SegmentCoreReaders.decRef(SegmentCoreReaders.java:157)
	at org.apache.lucene.index.SegmentReader.doClose(SegmentReader.java:175)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.ReadersAndUpdates.dropReaders(ReadersAndUpdates.java:182)
	at org.apache.lucene.index.IndexWriter$ReaderPool.dropAll(IndexWriter.java:603)
	at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2075)
	at org.apache.lucene.index.IndexWriter.rollback(IndexWriter.java:2033)
	at org.elasticsearch.index.engine.InternalEngine.closeNoLock(InternalEngine.java:993)
	at org.elasticsearch.index.engine.Engine.failEngine(Engine.java:540)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:685)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:615)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1255)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[2015-12-23 11:47:49,949][WARN ][index.engine             ] [data-nd2] [logstash-2015.12.23-v1][1] failed engine [refresh failed]
java.lang.NullPointerException
	at org.elasticsearch.indices.cache.query.IndicesQueryCache$1.onDocIdSetEviction(IndicesQueryCache.java:158)
	at org.apache.lucene.search.LRUQueryCache.clearCoreCacheKey(LRUQueryCache.java:313)
	at org.apache.lucene.search.LRUQueryCache$1.onClose(LRUQueryCache.java:276)
	at org.apache.lucene.index.SegmentCoreReaders.notifyCoreClosedListeners(SegmentCoreReaders.java:168)
	at org.apache.lucene.index.SegmentCoreReaders.decRef(SegmentCoreReaders.java:157)
	at org.apache.lucene.index.SegmentReader.doClose(SegmentReader.java:175)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.StandardDirectoryReader.doClose(StandardDirectoryReader.java:359)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.IndexReader.close(IndexReader.java:403)
	at org.apache.lucene.index.FilterDirectoryReader.doClose(FilterDirectoryReader.java:134)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:130)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:58)
	at org.apache.lucene.search.ReferenceManager.release(ReferenceManager.java:274)
	at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:189)
	at org.apache.lucene.search.ReferenceManager.maybeRefreshBlocking(ReferenceManager.java:253)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:678)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:615)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1255)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[2015-12-23 11:47:49,969][WARN ][index.shard              ] [data-nd2] [logstash-2015.12.23-v1][1] Failed to perform scheduled engine refresh
[logstash-2015.12.23-v1][[logstash-2015.12.23-v1][1]] RefreshFailedEngineException[Refresh failed]; nested: NullPointerException;
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:686)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:615)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1255)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.elasticsearch.indices.cache.query.IndicesQueryCache$1.onDocIdSetEviction(IndicesQueryCache.java:158)
	at org.apache.lucene.search.LRUQueryCache.clearCoreCacheKey(LRUQueryCache.java:313)
	at org.apache.lucene.search.LRUQueryCache$1.onClose(LRUQueryCache.java:276)
	at org.apache.lucene.index.SegmentCoreReaders.notifyCoreClosedListeners(SegmentCoreReaders.java:168)
	at org.apache.lucene.index.SegmentCoreReaders.decRef(SegmentCoreReaders.java:157)
	at org.apache.lucene.index.SegmentReader.doClose(SegmentReader.java:175)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.StandardDirectoryReader.doClose(StandardDirectoryReader.java:359)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.IndexReader.close(IndexReader.java:403)
	at org.apache.lucene.index.FilterDirectoryReader.doClose(FilterDirectoryReader.java:134)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:130)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:58)
	at org.apache.lucene.search.ReferenceManager.release(ReferenceManager.java:274)
	at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:189)
	at org.apache.lucene.search.ReferenceManager.maybeRefreshBlocking(ReferenceManager.java:253)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:678)
	... 5 more
[2015-12-23 11:47:49,989][WARN ][indices.cluster          ] [data-nd2] [[logstash-2015.12.23-v1][1]] marking and sending shard failed due to [engine failure, reason [refresh failed]]
java.lang.NullPointerException
	at org.elasticsearch.indices.cache.query.IndicesQueryCache$1.onDocIdSetEviction(IndicesQueryCache.java:158)
	at org.apache.lucene.search.LRUQueryCache.clearCoreCacheKey(LRUQueryCache.java:313)
	at org.apache.lucene.search.LRUQueryCache$1.onClose(LRUQueryCache.java:276)
	at org.apache.lucene.index.SegmentCoreReaders.notifyCoreClosedListeners(SegmentCoreReaders.java:168)
	at org.apache.lucene.index.SegmentCoreReaders.decRef(SegmentCoreReaders.java:157)
	at org.apache.lucene.index.SegmentReader.doClose(SegmentReader.java:175)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.StandardDirectoryReader.doClose(StandardDirectoryReader.java:359)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.index.IndexReader.close(IndexReader.java:403)
	at org.apache.lucene.index.FilterDirectoryReader.doClose(FilterDirectoryReader.java:134)
	at org.apache.lucene.index.IndexReader.decRef(IndexReader.java:253)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:130)
	at org.apache.lucene.search.SearcherManager.decRef(SearcherManager.java:58)
	at org.apache.lucene.search.ReferenceManager.release(ReferenceManager.java:274)
	at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:189)
	at org.apache.lucene.search.ReferenceManager.maybeRefreshBlocking(ReferenceManager.java:253)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:678)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:615)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1255)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by s1monw** [comment]
great, that helps a lot

**Comment by jpountz** [comment]
Fixed in 2.1.1 #15202
